### PR TITLE
Fix map.fitBounds() with non-zero bearing

### PR DESCRIPTION
**Pull request by TannerPerrien**
Fix bounds/bearing calculation bug (#10064) in the following methods:

* `map.fitBounds()`
* `map.fitScreenCoordinates()`
* `map.cameraForBounds()`

Consider bearings of 45 and 135 and how all four corners need to be taken into account.

Changes in tests:

* Fix tests for camera => #cameraForBounds (they were incorrect)
* Add new test for camera => #fitBounds with non-zero bearing
* Changed input coordinates in tests camera => #fitScreenCoordinates to make the input region not square

Before
before
After
after

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] include before/after visuals or gifs if this PR includes visual changes
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog>Fix bounds calculation where map.fitBounds(), map.fitScreenCoordinates(), and map.cameraForBounds() behave incorrectly with non-zero bearing. (#10064)</changelog>`

### CODE REVIEW

**Review by SnailBones — CHANGES_REQUESTED** [review]
Thanks for your work on this @TannerPerrien! I tested this out and it seems to work.

I did add a few small suggestions for code styling.

File: src/ui/camera.js [hunk 1]
```diff
@@ -621,15 +621,28 @@ class Camera extends Evented {
...
-        const p1world = tr.project(LngLat.convert(p1));
+        // We want to calculate the corners of the box defined by p0 and p1 in a coordinate system
+        // rotated to match the destination bearing. All four corners of the box must be taken
+        // into account because of camera rotation.
+        const p0LatLng = LngLat.convert(p0);
+        const p1LatLng = LngLat.convert(p1);
+        const p0world = tr.project(p0LatLng);
+        const p1world = tr.project(p1LatLng);
+        const p2world = tr.project(new LngLat(p0LatLng.lng, p1LatLng.lat));
+        const p3world = tr.project(new LngLat(p1LatLng.lng, p0LatLng.lat));
```
[the remark this answers is not in the document]
Comment (TannerPerrien) [reply src/ui/camera.js:632]: Much better. Thanks!

File: src/ui/camera.js [hunk 2]
```diff
@@ -621,15 +621,28 @@ class Camera extends Evented {
...
         const p0rotated = p0world.rotate(-degToRad(bearing));
         const p1rotated = p1world.rotate(-degToRad(bearing));
-
-        const upperRight = new Point(Math.max(p0rotated.x, p1rotated.x), Math.max(p0rotated.y, p1rotated.y));
-        const lowerLeft = new Point(Math.min(p0rotated.x, p1rotated.x), Math.min(p0rotated.y, p1rotated.y));
+        const p2rotated = p2world.rotate(-bearing * Math.PI / 180);
+        const p3rotated = p3world.rotate(-bearing * Math.PI / 180);
```
[the remark this answers is not in the document]
Comment (TannerPerrien) [reply src/ui/camera.js:636]: Done!

File: src/ui/camera.js [hunk 3]
```diff
@@ -621,15 +621,26 @@ class Camera extends Evented {
...
         const p1world = tr.project(LngLat.convert(p1));
+        const p2world = new Point(p0world.x, p1world.y);
+        const p3world = new Point(p1world.x, p0world.y);
         const p0rotated = p0world.rotate(-degToRad(bearing));
         const p1rotated = p1world.rotate(-degToRad(bearing));
-
-        const upperRight = new Point(Math.max(p0rotated.x, p1rotated.x), Math.max(p0rotated.y, p1rotated.y));
-        const lowerLeft = new Point(Math.min(p0rotated.x, p1rotated.x), Math.min(p0rotated.y, p1rotated.y));
+        const p2rotated = p2world.rotate(-degToRad(bearing));
+        const p3rotated = p3world.rotate(-degToRad(bearing));
```
[the remark this answers is not in the document]
Comment (TannerPerrien) [reply src/ui/camera.js]: Good call. Apologies for the slop!
